### ISSUE TITLE
Validate schema.Reference when embedded in nested FieldValidators

### ISSUE THREAD

**Issue by omani** [rs/rest-layer#57]
Hi,

I have the following schema:
```go
var Roles = schema.Schema{
	Fields: schema.Fields{
		"id":      schema.IDField,
		"created": schema.CreatedField,
		"updated": schema.UpdatedField,

		// fields
		"name": {
			Required:   true,
			Filterable: true,
			Sortable:   true,
			Validator: &schema.String{
				MinLen: 1,
				MaxLen: 63,
			},
		},
		"description": {
			Filterable: true,
			Sortable:   true,
			Validator: &schema.String{
				MaxLen: 63,
			},
		},
		"usernames": {
			Validator: &schema.Reference{
				Path: "users",
			},
		}
}
```

I would like to have usernames as an array of references. but I want to be able to pass the user's name instead of its IDField. something like:

```
pseudo POST request:
...
"usernames": ["omani", "rs"]
...
```

how do I do this? I guess there is no support for this in the Validator and I have to do this in a hook?

**Comment by omani** [comment]
maybe it is easier to use the `users` schema as a subresource to `usernames`. does this make more sense?

afaiu this results in having nested documents in mongodb instead of references, right?

**Comment by omani** [comment]
I've implemented a MutlGetByField method to be able to fetch for a subset of paths.

If anyone is interested you will find it [here](https://github.com/omani/rest-layer/commit/822fdef3ff40c41aaa72cdb0b632c599a3c27afa)

I still have to figure out how to have arrays of references though.

**Comment by omani** [comment]
Also see my latest commit [here](https://github.com/omani/rest-layer/commit/db42cb0fd39864cfe958fbec5a4adc31d919c280) which provides a new schema calles `ReferenceArray`.

**Comment by smyrman** [comment]
> Also see my latest commit here which provides a new schema calles ReferenceArray.

As for the `ReferenceArray` type, would not a normal `Array` validator with `ValuesValidator` set to `Reference` do the trick?
```go
...
"usernames": {
            Validator: &schema.Array{
                    ValuesValidator: &schema.Reference{
                            Path: "users",
                    },
            },
},
...
```
Or am I missing something?

**Comment by omani** [comment]
I thought about that also. but with that example the referencing does not work. That's why I came up with the ReferenceArray Validator.

**Comment by smyrman** [comment]
> I thought about that also. but with that example the referencing does not work
Sounds like a bug to me... @rs, would you agree?

As for the [first change](https://github.com/omani/rest-layer/commit/822fdef3ff40c41aaa72cdb0b632c599a3c27afa), I find them interesting, but they include breaking changes to the API. As of yet, rest-layer is not versioned, so that could result in breaking many people's code if mainlined. For a PR, I would assume it might be nicer not to rename `Get()` - > `GetByID()`, but I will let @rs confirm that...

@rs, would it also make sense to tag a `v0.x` relase for rest-later.io? A git tag would be good enough for most Go pacakge tools, e.g. [Glide](https://glide.sh/). Following [semver](http://semver.org/) conventions would be very nice.

**Comment by smyrman** [comment]
Alternatively to use get-by-field btw, you _could_ just make the userID be the ID field of the user. Useful if you also would want the username to be the lookup field in the API, and you don't need to allow changing it.

Another question to @rs, is if there is any way we could allow end-users to add custom Reference types. Would it e.g. be possible to let the reference logic in the `resource` package depen on optional validator interfaces being implemented, rather than hard-coding the type-check?

**Comment by rs** [comment]
An array does not have a sub-schema, that's why it doesn't check references inside it. It would be indeed better to support references in normal array instead of introducing a new validation. In all cases, this reference checking system is hacky. I'm not sure how to make it better but I'm quite not satisfied with it.

I would love to have this custom code in the reference code itself but I didn't find a clean way to do it without making `schema` depend on `rest` package.

**Comment by smyrman** [comment]
> I would love to have this custom code in the reference code itself but I didn't find a clean way to do it without making schema depend on rest package.

From https://github.com/rs/rest-layer/blob/master/rest/util.go#L186, I see you only use `Index.GetResorce(..)`, and then `Resource.Get(..)` to validate the reference. Perhaps there is a way using interfaces in the schema package, but it's indeed still a question if it can be done clean...

From the top of my head, combining `Index.GetResorce(..)` and `Resource.Get(..)` to one method on index (or a private wrapper to the index), would allow for a `ResorceItemGetter` interface in the schema package that could be used as a parameter to an additional validator interface:

```go
type ReferenceValidator interface {
        ValidateReference(i ResourceItemGetter, v interface{})
}
```

That might work, but is it clean? Not at all :-( Maybe it can inspire someone to get a better idea though:-P

**Comment by omani** [comment]
I would like to push this issue. Can we agree on something? Then we could start with the implementation.

**Comment by smyrman** [comment]
If you look at the inline comments in the first change, there are some advice from @rs.

As for the second change, I also think supporting Reference within a normal Array would be the best, if that's possible. 

As for cleaning up the Reference implementation as a whole and allow custom references, that's a separate issue.

**Comment by smyrman** [comment]
Please confirm that with @rs btw, I am just a contributor, like you.

**Comment by rs** [comment]
Yes I would prefer a fixed Array support for ref.

**Comment by omani** [comment]
did anybody start with an implementation for this? I want to prioritize this as I currently need this. so if you have some ideas when and where to start, feel free to post it. I will start to have a look at this today.

**Comment by smyrman** [comment]
> did anybody start with an implementation for this?
Don't think anyone is working on this.

**Comment by rs** [comment]
I tried to quick fix it and realized it was much more complex than it appeared so I postponed it for when I would have more time for it. Feel free to give it a try.

**Comment by smyrman** [comment]
Had a look at the code just now. It's _possible_ (but not pretty) to hard-code support for e.g.:
- Object
- Array with ValuesValidator == *(*schema.Reference ||  *schema.Object)
- Dict with ValuesValidator ==  (*schema.Reference ||  *schema.Object)
- AllOf ...
- AnyOf ...


It gets pretty tricky if you want to add support for custom deep types. 

Would be nice if we could think of a reasonable method that could allow Reference to be validated following the rest of the validation flow. I don't see how we could do that yet. Might introduce breaking changes.

**Comment by smyrman** [comment]
After sleeping on it, I found a plausible solution that might be good enough:

Both `schema.Reference`, _and_ any nested types like `schema.Schema`, `schema.Object`, `schema.Array`, `schema.Dict`, `schema.AllOf` and `schema.AnyOf` could be made to implement a new (optional) interface:

```go
type ReferenceValidator interface{
        // Validate references (recursively), or return on the first error.
        ValidateReferences(r ResourceItemChecker, value interface{}) error
}
```

The `ResourceItemChecker` would be a _wrapped_ version of the `resource.Index` and `context.Context` that implements the following interface:
```go
type ResourceItemChecker interface {
      ItemExists(path string, id interface{}) bool
}
```

Finally, we remove the `checkReferences` utility, and run `ValidateReferences` at the same place we run the normal `Validate` of the `Schema`.

This should solve:
- Move the validation of `schema.Reference` to happen within the `schema.Reference` Validate method.
- Allow `schema.Reference` instances to be placed in any nested field's Validator.
- Allow custom deep types to include references (as long as they implement the new optional interface).


@rs: Does this sound like a way forward to you? Am I missing something? I.e. do we need something to allow references to be embedded?

**Comment by rs** [comment]
On the paper it looks awesome. Would you have some time to work on a PoC?

**Comment by smyrman** [comment]
I have some time every now and then, so happy to have go at it. Can't promise I will be very fast.

**Comment by omani** [comment]
this would be awesome. feel free to make the start and code dump early so I can help you if you don't mind.

**Comment by smyrman** [comment]
@omani, @rs: I have been quite busy doing other stuff lately.

The linked https://github.com/smyrman/rest-layer/tree/breaking-compile is litterally just a code dump -- don't expect things to work, but please have a look and see what you think about the concept.

It is a slight modification to the original idea. Instead of creating another recursive ValidateXXX chain, I thought it would better to modify the Compile methods instead.

So I introduced the idea of a ReferenceCompiler that looks like this:
```go
// ReferenceCompiler is similar to the Compiler interface, but intended for types that implements, or may hold, a
// reference. All nested types must implement this interface.
type ReferenceCompiler interface {
	Compile(rl ResourceLookup) error
}

// ResourceLookup descries a type capable of checking if an item with a given ID exists for a given resource.
type ResourceLookup interface {
	ResourceExists(path string) bool
	ValidateID(path string, id interface{}) (interface{}, error)
}
```

I then started off replacing the implementation of Compiler with an implementation of ReferenceCompiler instead for Reference and nested types.

However, I discovered that the `compileDependences` appears quite broken from what I can see -- i.e. it doesn't really handle all those deep FieldValidators like we now want to do for `Reference`, but just the Schema.Schema (which we suggested deprecating in #77), and the usage within `schema.Object` appears to not even pass in the root schema which looks wrong.

To fix this, I see three alternatives:
1. Add more parameters to Compile for nested types (see below).
2. Add more Compile methods with different names for nested types. E.g.: Compile, ReferenceCompile, DependencyCompile

Solution 1, injects more crud....
```go
type DependencyCompiler interface {
	Compile(root *Schema) error
}
// Schema, Field and other deep FieldValidators implements this:
type DependencyAndReferenceCompiler interface {
	Compile(root *Schema, rl ResourceLookup) error
}
```
Another alternative would be to rename the Compile method in the interface, and let all nested types implement _several_ Compile methods that must all be called:

**Comment by smyrman** [comment]
As a FYI, I will not have the opportunity to write much code for the coming three weeks or so. Feel free to play with some code while I am away.

**Comment by smyrman** [comment]
@omani or @rs btw, do you mind updating the title to something that fits what we are trying to do now and remove the FEATURE PROVIDED tags?

E.g. "Validate schema.Reference when embedded in nested FieldValidators". If you don't like the rebranding, we could also create a new issue for this.

**Comment by smyrman** [comment]
> However, I discovered that the compileDependences appears quite broken from what I can see  -- i.e. it doesn't really handle all those deep FieldValidators like we now want to do for Reference, but just the Schema.Schema 

Oops - I think I misread the code. For some reason I thought the `compleDependencies` needed access to the _root_ schema. This probably makes the entire idea of embedding a root-schena into the Compile functions and/or context unnecessary. Must have been a bit tired.

**Comment by rs** [comment]
In your code example I don't understand what calls `reference.CompileReference` and how the `rl` property is initialized. It might be either to send a PR so we can comment in the code.

**Comment by smyrman** [comment]
@rs: You can comment in-line on the linked commit I believe. It's incomplete hack and not really ready for a PR. This is mearly a one afternoons quick hack that I did, and I did an early code dump as requested by @omani as I won't have time to work on this for a while.

After doing this hack, I became a bit unsure about if this was the right direction, so I cought up some quick reflections and alternate suggestions (including the ReferenceCompile method that tou mentions) in hope that it might be helpful for you or @omani if he was going to work more on it.

Answering from my phone.

**Comment by smyrman** [comment]
To answer directly on the `rl`, (sentdirectly  to `Compile` in the linked commit), because that would be part of either implementation. This is not yet implemented in the linked commit, but it should be trivial enough to implement a wrapper type to `resource.Index` that implements the `schema.ResourceLookup` interface. When calling resource compilation from the `resource.Index.Compile` method, we should be able to pass this wrapped index to the `Compile` method of each resource.

 When using the schema package alone, the user must provide his own ResourceLookup implementation and/or we could support nil values.

**Comment by smyrman** [comment]
@rs, Created a PR for the previouly linked commit so we can continue the discussion there.

**Comment by smyrman** [comment]
@omani, technically this _should_ be working now after #100. Can you report back if it still does not work.

**Comment by rs** [comment]
Please reopen if the problem is not fixed by #100.
